Sources/SwiftFusion/Core/Manifold.swift

```diff
@@ -0,0 +1,144 @@
+  ///   `localCoordinate(retract(b)) == b` for all `b \in B`.
+  @differentiable(wrt: local)
+  func retract(_ local: LocalCoordinate) -> Self
+
+  /// Inverse of `retract`.
+  ///
+  /// Satisfies the following properties:
+  /// - `localCoordinate(self) == LocalCoordinate.zero`
+  /// - There exists an open set `B` around `self` such that `localCoordinate(retract(b)) == b` for all
+  ///   `b \in B`.
+  @differentiable(wrt: global)
+  func localCoordinate(_ global: Self) -> LocalCoordinate
+}
+
+/// A point on a differentiable manifold.
+public protocol Manifold: Differentiable {
+  /// The manifold's global coordinate system.
+  associatedtype Coordinate: ManifoldCoordinate
+      where Coordinate.LocalCoordinate == Self.TangentVector
+
+  /// The coordinate of `self`.
+  ///
+  /// Note: The distinction between `coordinateStorage` and `coordinate` is a workaround until we
+  /// can define default derivatives for protocol requirements (TF-982). Until then, implementers
+  /// of this protocol must define `coordinateStorage`, and clients of this protocol must access
+  /// coordinate`. This allows us to define default derivatives for `coordinate` that translate
+  /// between the `ManifoldCoordinate` tangent space and the `Manifold` tangent space.
+  var coordinateStorage: Coordinate { get set }
+
+  /// Creates a manifold point with coordinate `coordinateStorage`.
+  ///
+  /// Note: The distinction between `init(coordinateStorage:)` and `init(coordinate:)` is a workaround until we
+  /// can define default derivatives for protocol requirements (TF-982). Until then, implementers
+  /// of this protocol must define `init(coordinateStorage:)`, and clients of this protocol must access
+  /// init(coordinate:)`. This allows us to define default derivatives for `init(coordinate:)` that translate
+  /// between the `ManifoldCoordinate` tangent space and the `Manifold` tangent space.
+  init(coordinateStorage: Coordinate)
+}
+
+extension Manifold {
+  /// The coordinate of `self`.
+  @differentiable
+  public var coordinate: Coordinate { coordinateStorage }
+
+  /// A custom derivative of `coordinate` that converts from the global coordinate system's
+  /// tangent vector to the local coordinate system's tangent vector, so that all functions on this
+  /// manifold using `coordinate` have derivatives involving local coordinates.
+  @derivative(of: coordinate)
+  @usableFromInline
+  func vjpCoordinate() -> (value: Coordinate, pullback: (Coordinate.TangentVector) -> TangentVector) {
+
+    // Explanation of this pullback:
+    //
+    // Let `f: Manifold -> Coordinate` be `f(x) = x.coordinateStorage`.
+    //
+    // `differential(at: x, in: f)` is a linear approximation of how changes in local
+    // coordinates around `x` lead to changes in global coordinates around `x.coordinateStorage`.
+    //
+    // `x.coordinateStorage.retract: LocalCoordinate -> Coordinate` defines _exactly_ how local
```

Comment:
I think you should say here `TangentVector -> Coordinate` to be consistent with pullback above, and also because I think within Manifold we should really only talk about Coordinate/TangentVector once we 'defined` them.